plugins/webp-uploads/settings.php

```diff
@@ -12,11 +12,28 @@
 }
 
 /**
- * Registers setting for generating both JPEG and WebP versions for image uploads.
+ * Registers setting for generating JPEG in addition to the selected modern format for image uploads.
  *
  * @since 1.0.0
+ * @since n.e.x.t The setting was made more general to cover outputting JPEG as a secondary type. The "webp" option naming
+ *                was left unchanged for backward compatibility.
+ * @since n.e.x.t The `perflab_modern_image_format` was added to enable selecting an output format.
+ *                Currently includes AVIF and WebP.
  */
 function webp_uploads_register_media_settings_field(): void {
+	register_setting(
+		'media',
+		'perflab_modern_image_format',
+		array(
+			'sanitize_callback' => static function ( $image_format ): string {
+				return in_array( $image_format, array( 'webp', 'avif' ), true ) ? $image_format : 'webp';
+			},
+			'type'              => 'string',
+			'default'           => 'avif',                                                                                                                                    // AVIF is the default if the editor supports it.
```

Comment:
Sites currently using the plugin and getting WebP, when they upgrade to the new version of the plugin will find that `avif` is the new default if they support AVIF, right? Is that OK?

Reply:
I think it is ok because it is a lab plugin, lets discuss tomorrow at the chat. 

We could also bump the plugin version to 2.0, since this is a pretty major new feature; this helps indicate potentially breaking changes.

Reply:
I think the change to the default option value should only apply if this option isn't already set. I assume that this is a new option, so is there any way we could tell if this feature was already in use, and if so, stick with the WebP option? For all new installs, defaulting to AVIF makes a lot of sense.

Additionally, in the sanitize callback, when an invalid option is passed, I am curious why we're falling back to WebP even though we're setting AVIF as the default. Is shouldn't we fallback to the same default?

Reply:
> I think the change to the default option value should only apply if this option isn't already set. I assume that this is a new option, so is there any way we could tell if this feature was already in use, and if so, stick with the WebP option? For all new installs, defaulting to AVIF makes a lot of sense.

I think the problem is that the option may not be set for existing sites. For example, there is a `perflab_generate_webp_and_jpeg` registered setting, but the option will only exist in the database if a user went to the Media screen and hit the Save Changes button.

> Additionally, in the sanitize callback, when an invalid option is passed, I am curious why we're falling back to WebP even though we're setting AVIF as the default. Is shouldn't we fallback to the same default?

Humm. Good point. I guess that should be made `avif` by default then, as well in `webp_uploads_get_image_output_format()`. If AVIF is not actually available, then it will force the use of WebP anyway.

Reply:
> I think the change to the default option value should only apply if this option isn't already set. I assume that this is a new option, so is there any way we could tell if this feature was already in use, and if so, stick with the WebP option? For all new installs, defaulting to AVIF makes a lot of sense.

Or maybe I misunderstood. This setting is new, yes. So you're saying that we should default to WebP if that setting is not set? But that setting will only get populated if the user goes to the settings screen and hits Save Changes, so in that case the default value wouldn't be used anyway.

Reply:
Honestly, I'm conflicted about what we should do. I think there's some validity in maintaining the WebP default for any sites that already had the plugin installed—particularly since originally plugin was marketed only as a WebP plugin. But I'm also sympathetic to the fact that this is a feature plugin intended to improve the performance of a site by serving modern image formats and changing the default format to the most performant option that the site supports would also make sense.

I guess if we had a way to determine that the plugin was previously installed prior to AVIF support being added, we could use that to keep the site on WebP and prompt the site owner to opt-in to AVIF, while setting AVIF as the default for any new versions of the plugin, but I'm not sure if there's a good way to make that determinations—or if we even should.

Reply:
Wouldn't bumping the version to 2.0 as well as adding an Upgrade Notice suffice?

Reply:
Yes. I think it we're changing the default to AVIF, which will essentially change the format being used by existing sites that are currently using WebP, then those two things are good to do. Can we add a link in the upgrade notice or add an admin pointer after the upgrade is installed to alert users that they can control the format in the settings?

Reply:
> Can we add a link in the upgrade notice or add an admin pointer after the upgrade is installed 

We could add a link to the upgrade notice, but would need to do that ahead of time in a separate release.

An admin pointer after the upgrade should work, I can try that.

Reply:
I thought the Upgrade Notice section of the readme got displayed somehow prior to upgrading:

https://github.com/WordPress/performance/blob/0a9570795dbc602cbc305d2b47b3a5a589f547cd/plugins/performance-lab/readme.txt#L485-L493

But I don't see it displayed anywhere that I can see when I'm downgraded to 2.9.0.

Reply:
Oh, I found it! The Web Stories plugin has an [upgrade notice](https://github.com/GoogleForCreators/web-stories-wp/blob/91460e88f8a2eecd0c8455f9a4f9b4a612677ef0/readme.txt#L141-L145) for the current version.

It gets displayed with the plugin on the Upgrade screen:

![image](https://github.com/WordPress/performance/assets/134745/a4f405db-4779-42a3-ba5e-8634604c6db5)

Honestly, I've never seen this before, and I rarely upgrade plugins from this screen now, as the upgrade from the plugins list table is a better experience. Still, maybe it is sufficient instead of adding an admin pointer?

Reply:
> Still, maybe it is sufficient instead of adding an admin pointer?

Yes, honestly I prefer this approach. I find the admin pointers distracting and most people probably dismiss them without even seeing what they say.